### ISSUE TITLE
Non-deterministic behaviour in deleting object in nested `TCollections`

### ISSUE THREAD

**Issue by mconcas** [root-project/root#14504]
### Check duplicate issues.

- [X] Checked for duplicates

### TL;DR:
Running multiple times the reproducer posted here I get inconsistent behaviour on Ubuntu 22.04.1.
It goes in `SIGSEGV` ~25% of the times, and would expect it to either fail consistently or always succeed.

### Context:
The `deleteTCollections()` function in the reproducer is a utility I copied verbatim from our framework.
The reproducer is a minimal version of a unit test we run in our CI.
We started to experience this regression (test failing) on our CI when we tried to move to a version of root compiled with C++20.
It turns out that this is reproduced in root v6-28 and v6-30 even if C++20 **is not enabled**.
It appeared only on Ubuntu tests, so far.

I am attaching the test.C asap in editing this issue.

### Reproducer

```C++
#include <TObject.h>
#include <TObjArray.h>
#include <TList.h>
#include <TH1I.h>
#include <TH2I.h>
#include <iostream>

void deleteTCollections(TObject *obj)
{
  std::cout << "Processing " << ((TNamed *)obj)->TNamed::GetName() << " of type " << obj->ClassName() << std::endl;
  if (auto c = dynamic_cast<TCollection *>(obj))
  {
    std::cout << "Entering TCollection... " << std::endl;
    c->SetOwner(false);
    auto iter = c->MakeIterator();
    while (auto element = iter->Next())
    {
      deleteTCollections(element);
    }
    delete iter;
    delete c;
  }
  else
  {
    delete obj;
  }
}

void test()
{
  int bins{10}, min{0}, max{10};
  TObjArray *main = new TObjArray();
  main->SetOwner(false);

  auto *collectionInside = new TList();
  collectionInside->SetOwner(true);

  TH1I *histo1DinsideCollection = new TH1I("histo 1d", "histo 1d", bins, min, max);
  histo1DinsideCollection->Fill(2);
  collectionInside->Add(histo1DinsideCollection);

  main->Add(collectionInside);
  std::cout << " ====== Listing elements in lists... ====== " << std::endl;
  for (int i = 0; i < main->GetEntries(); ++i)
  {
    std::cout << "main[" << i << "] = " << main->At(i) << " is a: " << main->At(i)->ClassName() << " with title: " << ((TNamed *)(main->At(i)))->GetTitle() << std::endl;
  }
  for (int i = 0; i < collectionInside->GetEntries(); ++i)
  {
    std::cout << "\t collectionInside[" << i << "] = " << collectionInside->At(i) << " is a: " << collectionInside->At(i)->ClassName() << " with title: " << ((TNamed *)(collectionInside->At(i)))->GetTitle() << std::endl;
  }
  std::cout << " ====== Deleting... ====== " << std::endl;
  deleteTCollections(main);
  return;
}
```

### ROOT version

```
ROOT v6.30/01
Built for linuxx8664gcc on Jan 30 2024, 16:24:12
From tags/v6-30-01-alice2@v6-25-02-6400-g526782cfe8
With c++ (GCC) 12.2.0
Binary directory: /ssd_data/builds/latest/sw/ubuntu2204_x86-64/ROOT/526782cfe8-local1/bin
```

### Installation method

aliBuild

### Operating system

Ubuntu 22.04.01

### Additional context

```bash
$> uname -ra
Linux <hostname> 6.2.0-39-generic #40~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Thu Nov 16 10:53:04 UTC 2 x86_64 x86_64 x86_64 GNU/Linux
$> lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 22.04.3 LTS
Release:	22.04
Codename:	jammy
```

[test.txt](https://github.com/root-project/root/files/14112818/test.txt)

CC: @ktf

**Comment by ferdymercury** [comment]
If you run it with:

`valgrind --leak-check=full --suppressions=$ROOTSYS/etc/valgrind-root.supp root.exe -l -b -q test.cpp+`

you'll get:

```
==721499== Invalid read of size 4
==721499==    at 0x4AF2199: TList::Clear(char const*) (in /opt/root/lib/libCore.so.6.30.04)
==721499==    by 0x4AF26D4: TList::~TList() (in /opt/root/lib/libCore.so.6.30.04)
==721499==    by 0x4AF287C: TList::~TList() (in /opt/root/lib/libCore.so.6.30.04)
==721499==    by 0x2729A8D9: deleteTCollections(TObject*) (in /tmp/tmp/test_cpp.so)
==721499==    by 0x2729F039: ???
==721499==    by 0x6AA94F7: cling::IncrementalExecutor::executeWrapper(llvm::StringRef, cling::Value*) const (in /opt/root/lib/libCling.so.6.30.04)
==721499==    by 0x6A31E4B: cling::Interpreter::RunFunction(clang::FunctionDecl const*, cling::Value*) (in /opt/root/lib/libCling.so.6.30.04)
==721499==    by 0x6A32596: cling::Interpreter::EvaluateInternal(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, cling::CompilationOptions, cling::Value*, cling::Transaction**, unsigned long) (in /opt/root/lib/libCling.so.6.30.04)
==721499==    by 0x6A327B7: cling::Interpreter::process(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, cling::Value*, cling::Transaction**, bool) (in /opt/root/lib/libCling.so.6.30.04)
==721499==    by 0x6B09BC6: cling::MetaProcessor::process(llvm::StringRef, cling::Interpreter::CompilationResult&, cling::Value*, bool) (in /opt/root/lib/libCling.so.6.30.04)
==721499==    by 0x693813B: HandleInterpreterException(cling::MetaProcessor*, char const*, cling::Interpreter::CompilationResult&, cling::Value*) (in /opt/root/lib/libCling.so.6.30.04)
==721499==    by 0x694C6E7: TCling::ProcessLine(char const*, TInterpreter::EErrorCode*) (in /opt/root/lib/libCling.so.6.30.04)
==721499==  Address 0xf0ff6dc is 12 bytes inside a block of size 1,000 free'd
==721499==    at 0x484B8AF: operator delete(void*) (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==721499==    by 0x2729A8D9: deleteTCollections(TObject*) (in /tmp/tmp/test_cpp.so)
==721499==    by 0x2729A8D9: deleteTCollections(TObject*) (in /tmp/tmp/test_cpp.so)
==721499==    by 0x2729F039: ???
==721499==    by 0x6AA94F7: cling::IncrementalExecutor::executeWrapper(llvm::StringRef, cling::Value*) const (in /opt/root/lib/libCling.so.6.30.04)
==721499==    by 0x6A31E4B: cling::Interpreter::RunFunction(clang::FunctionDecl const*, cling::Value*) (in /opt/root/lib/libCling.so.6.30.04)
==721499==    by 0x6A32596: cling::Interpreter::EvaluateInternal(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, cling::CompilationOptions, cling::Value*, cling::Transaction**, unsigned long) (in /opt/root/lib/libCling.so.6.30.04)
==721499==    by 0x6A327B7: cling::Interpreter::process(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, cling::Value*, cling::Transaction**, bool) (in /opt/root/lib/libCling.so.6.30.04)
==721499==    by 0x6B09BC6: cling::MetaProcessor::process(llvm::StringRef, cling::Interpreter::CompilationResult&, cling::Value*, bool) (in /opt/root/lib/libCling.so.6.30.04)
==721499==    by 0x693813B: HandleInterpreterException(cling::MetaProcessor*, char const*, cling::Interpreter::CompilationResult&, cling::Value*) (in /opt/root/lib/libCling.so.6.30.04)
==721499==    by 0x694C6E7: TCling::ProcessLine(char const*, TInterpreter::EErrorCode*) (in /opt/root/lib/libCling.so.6.30.04)
==721499==    by 0x694CB7A: TCling::ProcessLineSynch(char const*, TInterpreter::EErrorCode*) (in /opt/root/lib/libCling.so.6.30.04)
==721499==  Block was alloc'd at
==721499==    at 0x4849013: operator new(unsigned long) (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==721499==    by 0x4A95EDD: TStorage::ObjectAlloc(unsigned long) (in /opt/root/lib/libCore.so.6.30.04)
==721499==    by 0x2729AAAE: test() (in /tmp/tmp/test_cpp.so)
==721499==    by 0x2729F039: ???
==721499==    by 0x6AA94F7: cling::IncrementalExecutor::executeWrapper(llvm::StringRef, cling::Value*) const (in /opt/root/lib/libCling.so.6.30.04)
==721499==    by 0x6A31E4B: cling::Interpreter::RunFunction(clang::FunctionDecl const*, cling::Value*) (in /opt/root/lib/libCling.so.6.30.04)
==721499==    by 0x6A32596: cling::Interpreter::EvaluateInternal(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, cling::CompilationOptions, cling::Value*, cling::Transaction**, unsigned long) (in /opt/root/lib/libCling.so.6.30.04)
==721499==    by 0x6A327B7: cling::Interpreter::process(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, cling::Value*, cling::Transaction**, bool) (in /opt/root/lib/libCling.so.6.30.04)
==721499==    by 0x6B09BC6: cling::MetaProcessor::process(llvm::StringRef, cling::Interpreter::CompilationResult&, cling::Value*, bool) (in /opt/root/lib/libCling.so.6.30.04)
==721499==    by 0x693813B: HandleInterpreterException(cling::MetaProcessor*, char const*, cling::Interpreter::CompilationResult&, cling::Value*) (in /opt/root/lib/libCling.so.6.30.04)
==721499==    by 0x694C6E7: TCling::ProcessLine(char const*, TInterpreter::EErrorCode*) (in /opt/root/lib/libCling.so.6.30.04)
==721499==    by 0x694CB7A: TCling::ProcessLineSynch(char const*, TInterpreter::EErrorCode*) (in /opt/root/lib/libCling.so.6.30.04)
==721499== 
Error in <TList::Clear>: A list is accessing an object (0xf0ff6d0) already deleted (list name = TList)
```

**Comment by mconcas** [comment]
I conclude that manually deleting the objects is not ok, even if the list is not the owner of the objects.
Or shall I also mark the objects I want to manually delete as `nodelete` in the TList?

Am I missing any good practice here?

**Comment by pcanal** [comment]
If  you are deleting an object that is contained in a `TCollection`, you must remove it from that collection, otherwise the collection has a dangling pointer (that it will use during certain operations, including `Clear` but could be any other operation, etc `ls()` ).

**Comment by ferdymercury** [comment]
Maybe setting the content of the dangling pointer to NULL just after delete helps ?

**Comment by mconcas** [comment]
> If you are deleting an object that is contained in a `TCollection`, you must remove it from that collection, otherwise the collection has a dangling pointer (that it will use during certain operations, including `Clear` but could be any other operation, etc `ls()` ).

Ok, that makes sense, and indeed, it is impossible to do the other way around (i.e. `delete obj` before and `Remove(obj)` from the list after) as there is no obvious way to know if the object has been deleted. So, there is little you can do from the ROOT side.

> Maybe setting the content of the dangling TList pointer to NULL just after delete helps ?

It would break the recursive structure of the `deleteTCollections` function, which we'll need to rethink. 

Thanks for the clarifications and for your time.

**Comment by wdconinc** [comment]
We are seeing similar issues as a regression when writing ROOT files with 6.30 vs 6.28 (using AIDASoft/podio), and then opening them in older ROOT versions. We're still narrowing this into a reproducer but it seems like something changed in this context to make this issue appear now, with 6.30...

File written by 6.30, read by 6.24
![York_6p30_File-gb9kj5ui57yffrq1qwzj75ymah.png](https://github.com/root-project/root/assets/4656391/d33a096d-1b42-4dc4-b9e9-fcfdb40bac07)

File written by 6.28, read by 6.24
![York_6p28_File-5c8tqx5ofp8tjdo7fk1518r6oy.png](https://github.com/root-project/root/assets/4656391/d9f89ac4-d4e8-4684-9679-5d3491f0f2d5)

**Comment by wdconinc** [comment]
(I'll file a new issue later today and refer to this one. Consider this a heads-up :-) )

**Comment by chrisburr** [comment]
@wdconinc did you open this issue anywhere? I've also ran what appears to be the same issue.

**Comment by wdconinc** [comment]
No, unfortunately I haven't found the time yet for that one :-) It's been hard to write a reproducer that isn't "take our entire software stack and run a full simulation".
